contracts/examples/crypto-kitties/kitty-auction/src/lib.rs

```diff
@@ -0,0 +1,334 @@
+
+		// refund losing bid
+		if auction.current_winner != Address::zero() {
+			self.send_tx(&auction.current_winner, &auction.current_bid, "bid refund");
+		}
+
+		// update auction bid and winner
+		auction.current_bid = payment;
+		auction.current_winner = caller;
+		self.set_auction(kitty_id, &auction);
+
+		Ok(())
+	}
+
+	#[endpoint(endAuction)]
+	fn end_auction(&self, kitty_id: u32) -> SCResult<()> {
+		require!(
+			self.is_up_for_auction(kitty_id),
+			"kitty is not up for auction!"
+		);
+
+		let auction = self.get_auction(kitty_id);
+
+		require!(
+			self.get_block_timestamp() >= auction.deadline
+				|| auction.current_bid == auction.ending_price,
+			"auction has not ended yet!"
+		);
+
+		if auction.current_winner != Address::zero() {
+			match auction.auction_type {
+				AuctionType::Selling => {
+					self._transfer_to(auction.current_winner, kitty_id);
+				},
+				AuctionType::Siring => {
+					self._approve_siring(auction.current_winner, kitty_id);
+				},
+			}
+		} else {
+			// no bids were made
+			self.clear_auction(kitty_id);
+		}
+
+		Ok(())
+	}
+
+	// callbacks
+
+	#[callback]
+	fn allow_auctioning_callback(
+		&self,
+		result: AsyncCallResult<()>,
+		#[callback_arg] auction_type: AuctionType,
+		#[callback_arg] cb_kitty_id: u32,
+		#[callback_arg] starting_price: BigUint,
+		#[callback_arg] ending_price: BigUint,
+		#[callback_arg] deadline: u64,
+		#[callback_arg] kitty_owner: Address,
+	) {
+		match result {
+			AsyncCallResult::Ok(()) => {
+				let auction = Auction::new(
+					auction_type,
+					&starting_price,
+					&ending_price,
+					deadline,
+					&kitty_owner,
+				);
+
+				self.set_auction(cb_kitty_id, &auction);
+			},
+			AsyncCallResult::Err(_) => {
+				// nothing to revert in case of error
+			},
+		}
+	}
+
+	#[callback]
+	fn transfer_callback(&self, result: AsyncCallResult<()>, #[callback_arg] cb_kitty_id: u32) {
+		match result {
+			AsyncCallResult::Ok(()) => {
+				let auction = self.get_auction(cb_kitty_id);
+				self.clear_auction(cb_kitty_id);
+
+				// send winning bid money to kitty owner
+				self.send_tx(&auction.kitty_owner, &auction.current_bid, "sold kitty");
+			},
+			AsyncCallResult::Err(_) => {
+				// this can only fail if the kitty_ownership contract address is invalid
+				// nothing to revert in case of error
+			},
+		}
+	}
+
+	#[callback]
+	fn approve_siring_callback(
+		&self,
+		result: AsyncCallResult<()>,
+		#[callback_arg] cb_kitty_id: u32,
+	) {
+		match result {
+			AsyncCallResult::Ok(()) => {
+				let auction = self.get_auction(cb_kitty_id);
+
+				// transfer kitty back to its owner
+				self._transfer_to(auction.kitty_owner, cb_kitty_id);
+
+				// auction data will be cleared in the transfer callback
+				// winning bid money will be sent as well
+			},
+			AsyncCallResult::Err(_) => {
+				// this can only fail if the kitty_ownership contract address is invalid
+				// nothing to revert in case of error
+			},
+		}
+	}
+
+	// private
+
+	fn _create_auction(
```

Comment:
no need for leading underscores anymore

Reply:
I prefer having underscores for private functions, but I'm willing to change that if it's not a good convention.